util/conversion/conversion.go

```diff
@@ -108,23 +108,27 @@ func MarshalData(src metav1.Object, dst metav1.Object) error {
 	if err != nil {
 		return err
 	}
+	annotations := dst.GetAnnotations()
 	if dst.GetAnnotations() == nil {
-		dst.SetAnnotations(map[string]string{})
+		annotations = map[string]string{}
 	}
-	dst.GetAnnotations()[DataAnnotation] = string(data)
+	annotations[DataAnnotation] = string(data)
+	dst.SetAnnotations(annotations)
```

Comment:
do we need this? if annotations is a map wouldn't it get updated as it's a reference?

Reply:
I feel like this is the exact same bug as https://github.com/kubernetes-sigs/cluster-api/pull/4373/files#diff-e25fb86695a8022fa1611100bd244f9b4185f28629d19830e83d18b4fa2ce710R54

we only need to set it if it was nil before

Reply:
It works for normal objects, although for unstructured when it gets annotations, it actually does a copy https://github.com/kubernetes/kubernetes/blob/master/staging/src/k8s.io/apimachinery/pkg/apis/meta/v1/unstructured/unstructured.go#L407-L410 and https://github.com/kubernetes/kubernetes/blob/e8760b95bb0a0dfb5c184fd4e5ab95da3650a040/staging/src/k8s.io/apimachinery/pkg/apis/meta/v1/unstructured/helpers.go#L169